pkg/deploy/controller/deploymentconfig/factory.go

```diff
@@ -9,84 +9,232 @@ import (
-		ListFunc: func(options kapi.ListOptions) (runtime.Object, error) {
-			return factory.Client.DeploymentConfigs(kapi.NamespaceAll).List(options)
-		},
-		WatchFunc: func(options kapi.ListOptions) (watch.Interface, error) {
-			return factory.Client.DeploymentConfigs(kapi.NamespaceAll).Watch(options)
-		},
+// Run begins watching and syncing.
+func (c *DeploymentConfigController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+
+	// Wait for the rc and pod stores to sync before starting any work in this controller.
+	ready := make(chan struct{})
+	go c.waitForSyncedStore(ready)
+	select {
+	case <-ready:
+	case <-stopCh:
+		return
 	}
-	queue := cache.NewFIFO(cache.MetaNamespaceKeyFunc)
-	cache.NewReflector(deploymentConfigLW, &deployapi.DeploymentConfig{}, queue, 2*time.Minute).Run()
 
-	eventBroadcaster := record.NewBroadcaster()
-	eventBroadcaster.StartRecordingToSink(factory.KubeClient.Events(""))
-	recorder := eventBroadcaster.NewRecorder(kapi.EventSource{Component: "deploymentconfig-controller"})
+	for i := 0; i < workers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+
+	<-stopCh
+	glog.Infof("Shutting down deploymentconfig controller")
+	c.queue.ShutDown()
+}
+
+func (c *DeploymentConfigController) waitForSyncedStore(ready chan struct{}) {
+	defer utilruntime.HandleCrash()
+
+	for !c.rcStoreSynced() {
+		glog.V(4).Infof("Waiting for the rc controller to sync before starting the deployment config controller workers")
+		time.Sleep(StoreSyncedPollPeriod)
+	}
+	close(ready)
+}
+
+func (c *DeploymentConfigController) addDeploymentConfig(obj interface{}) {
+	dc := obj.(*deployapi.DeploymentConfig)
+	glog.V(4).Infof("Adding deployment config %q", dc.Name)
+	c.enqueueDeploymentConfig(dc)
+}
+
+func (c *DeploymentConfigController) updateDeploymentConfig(old, cur interface{}) {
+	oldDc := old.(*deployapi.DeploymentConfig)
+	glog.V(4).Infof("Updating deployment config %q", oldDc.Name)
+	c.enqueueDeploymentConfig(cur.(*deployapi.DeploymentConfig))
+}
+
+func (c *DeploymentConfigController) deleteDeploymentConfig(obj interface{}) {
+	dc, ok := obj.(*deployapi.DeploymentConfig)
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			glog.Errorf("Couldn't get object from tombstone %+v", obj)
+			return
+		}
+		dc, ok = tombstone.Obj.(*deployapi.DeploymentConfig)
+		if !ok {
+			glog.Errorf("Tombstone contained object that is not a deployment config: %+v", obj)
+			return
+		}
+	}
+	glog.V(4).Infof("Deleting deployment config %q", dc.Name)
+	c.enqueueDeploymentConfig(dc)
+}
+
+// addReplicationController enqueues the deployment that manages a replicationcontroller when the replicationcontroller is created.
+func (c *DeploymentConfigController) addReplicationController(obj interface{}) {
+	rc := obj.(*kapi.ReplicationController)
+	glog.V(4).Infof("Replication controller %q added.", rc.Name)
+	if dc, err := c.dcStore.GetConfigForController(rc); err == nil && dc != nil {
```

Comment:
This code means that your sync function as to wait for DCs to sync too.  Even with the wait, there are pathological cases of highly latent watches, but a comment will do as long as you wait for a sync.